deploy/runtools/run_farm_deploy_managers.py

```diff
@@ -15,8 +15,8 @@
 from runtools.utils import has_sudo
 
 from typing import List, Dict, Optional, Union, TYPE_CHECKING
+from mypy_boto3_ec2.service_resource import Instance as EC2InstanceResource
 if TYPE_CHECKING:
```

Comment:
When initially doing the type-checking work, I often hit many circular imports (IIRC around the `runtools.firesim_topology_elements` imports). I decided it's safer to add the `if` statement to make things explicitly type-checking related or not.

I've never tried `pyflakes`. If it does work, I think we could add it to CI.

Reply:
I think since you started doing the type-checking work, we've bumped versions of Python and also added `from __future__ import annotations` to all files.  I'll let you know what I find with pyflakes.

Reply:
Unfortunately `pyflakes` doesn't seem to catch the undefined symbol... It only complains about:
```
bash-4.2$ pyflakes deploy/runtools/run_farm_deploy_managers.py 
deploy/runtools/run_farm_deploy_managers.py:11:1: 'time' imported but unused
deploy/runtools/run_farm_deploy_managers.py:20:5: 'runtools.firesim_topology_elements.FireSimSwitchNode' imported but unused
deploy/runtools/run_farm_deploy_managers.py:20:5: 'runtools.firesim_topology_elements.FireSimServerNode' imported but unused
deploy/runtools/run_farm_deploy_managers.py:628:17: local variable 'collect' is assigned to but never used
```

Reply:
`pylint` is full of all kinds of suggestions but also doesn't flag the undefined symbol error.

<details>
<summary> pylint output </summary>

```
bash-4.2$ pylint deploy/runtools/run_farm_deploy_managers.py 
************* Module runtools.run_farm_deploy_managers
deploy/runtools/run_farm_deploy_managers.py:56:0: C0301: Line too long (101/100) (line-too-long)
deploy/runtools/run_farm_deploy_managers.py:154:0: C0301: Line too long (118/100) (line-too-long)
deploy/runtools/run_farm_deploy_managers.py:165:0: C0301: Line too long (116/100) (line-too-long)
deploy/runtools/run_farm_deploy_managers.py:175:0: C0301: Line too long (120/100) (line-too-long)
deploy/runtools/run_farm_deploy_managers.py:190:0: C0301: Line too long (106/100) (line-too-long)
deploy/runtools/run_farm_deploy_managers.py:201:0: C0301: Line too long (104/100) (line-too-long)
deploy/runtools/run_farm_deploy_managers.py:213:0: C0301: Line too long (104/100) (line-too-long)
deploy/runtools/run_farm_deploy_managers.py:225:0: C0301: Line too long (103/100) (line-too-long)
deploy/runtools/run_farm_deploy_managers.py:298:0: C0301: Line too long (107/100) (line-too-long)
deploy/runtools/run_farm_deploy_managers.py:440:0: C0301: Line too long (128/100) (line-too-long)
deploy/runtools/run_farm_deploy_managers.py:442:0: C0301: Line too long (108/100) (line-too-long)
deploy/runtools/run_farm_deploy_managers.py:443:0: C0301: Line too long (132/100) (line-too-long)
deploy/runtools/run_farm_deploy_managers.py:504:0: C0301: Line too long (130/100) (line-too-long)
deploy/runtools/run_farm_deploy_managers.py:514:0: C0301: Line too long (102/100) (line-too-long)
deploy/runtools/run_farm_deploy_managers.py:524:0: C0301: Line too long (105/100) (line-too-long)
deploy/runtools/run_farm_deploy_managers.py:525:0: C0301: Line too long (113/100) (line-too-long)
deploy/runtools/run_farm_deploy_managers.py:530:0: C0301: Line too long (114/100) (line-too-long)
deploy/runtools/run_farm_deploy_managers.py:531:0: C0301: Line too long (129/100) (line-too-long)
deploy/runtools/run_farm_deploy_managers.py:533:0: C0301: Line too long (105/100) (line-too-long)
deploy/runtools/run_farm_deploy_managers.py:534:0: C0301: Line too long (119/100) (line-too-long)
deploy/runtools/run_farm_deploy_managers.py:535:0: C0301: Line too long (129/100) (line-too-long)
deploy/runtools/run_farm_deploy_managers.py:556:0: C0301: Line too long (128/100) (line-too-long)
deploy/runtools/run_farm_deploy_managers.py:659:0: C0305: Trailing newlines (trailing-newlines)
deploy/runtools/run_farm_deploy_managers.py:111:13: W0511: ## XXX Centos Specific (fixme)
deploy/runtools/run_farm_deploy_managers.py:547:13: W0511: TODO: can make these values automatically be chosen based on link lat (fixme)
deploy/runtools/run_farm_deploy_managers.py:36:25: C0209: Formatting a regular string which could be a f-string (consider-using-f-string)
deploy/runtools/run_farm_deploy_managers.py:48:28: C0201: Consider iterating the dictionary directly instead of calling .keys() (consider-iterating-dictionary)
deploy/runtools/run_farm_deploy_managers.py:26:0: R0903: Too few public methods (1/2) (too-few-public-methods)
deploy/runtools/run_farm_deploy_managers.py:101:8: W1201: Use lazy % formatting in logging functions (logging-not-lazy)
deploy/runtools/run_farm_deploy_managers.py:101:24: C0209: Formatting a regular string which could be a f-string (consider-using-f-string)
deploy/runtools/run_farm_deploy_managers.py:122:16: C0209: Formatting a regular string which could be a f-string (consider-using-f-string)
deploy/runtools/run_farm_deploy_managers.py:144:35: C0209: Formatting a regular string which could be a f-string (consider-using-f-string)
deploy/runtools/run_farm_deploy_managers.py:158:29: C0209: Formatting a regular string which could be a f-string (consider-using-f-string)
deploy/runtools/run_farm_deploy_managers.py:160:16: C0209: Formatting a regular string which could be a f-string (consider-using-f-string)
deploy/runtools/run_farm_deploy_managers.py:170:16: C0209: Formatting a regular string which could be a f-string (consider-using-f-string)
deploy/runtools/run_farm_deploy_managers.py:175:33: C0209: Formatting a regular string which could be a f-string (consider-using-f-string)
deploy/runtools/run_farm_deploy_managers.py:177:32: C0209: Formatting a regular string which could be a f-string (consider-using-f-string)
deploy/runtools/run_farm_deploy_managers.py:178:16: C0209: Formatting a regular string which could be a f-string (consider-using-f-string)
deploy/runtools/run_farm_deploy_managers.py:190:33: C0209: Formatting a regular string which could be a f-string (consider-using-f-string)
deploy/runtools/run_farm_deploy_managers.py:192:32: C0209: Formatting a regular string which could be a f-string (consider-using-f-string)
deploy/runtools/run_farm_deploy_managers.py:195:12: E1129: Context manager 'generator' doesn't implement __enter__ and __exit__. (not-context-manager)
deploy/runtools/run_farm_deploy_managers.py:203:29: C0209: Formatting a regular string which could be a f-string (consider-using-f-string)
deploy/runtools/run_farm_deploy_managers.py:206:12: E1129: Context manager 'generator' doesn't implement __enter__ and __exit__. (not-context-manager)
deploy/runtools/run_farm_deploy_managers.py:213:33: C0209: Formatting a regular string which could be a f-string (consider-using-f-string)
deploy/runtools/run_farm_deploy_managers.py:280:8: E1129: Context manager 'generator' doesn't implement __enter__ and __exit__. (not-context-manager)
deploy/runtools/run_farm_deploy_managers.py:298:4: R0914: Too many local variables (20/15) (too-many-locals)
deploy/runtools/run_farm_deploy_managers.py:304:8: W1201: Use lazy % formatting in logging functions (logging-not-lazy)
deploy/runtools/run_farm_deploy_managers.py:315:16: C0200: Consider using enumerate instead of iterating with range and len (consider-using-enumerate)
deploy/runtools/run_farm_deploy_managers.py:325:36: R1735: Consider using {} instead of dict() (use-dict-literal)
deploy/runtools/run_farm_deploy_managers.py:325:52: R1735: Consider using {} instead of dict() (use-dict-literal)
deploy/runtools/run_farm_deploy_managers.py:333:63: R1735: Consider using {} instead of dict() (use-dict-literal)
deploy/runtools/run_farm_deploy_managers.py:342:12: W1201: Use lazy % formatting in logging functions (logging-not-lazy)
deploy/runtools/run_farm_deploy_managers.py:344:12: W1201: Use lazy % formatting in logging functions (logging-not-lazy)
deploy/runtools/run_farm_deploy_managers.py:345:27: R1729: Use a generator instead 'all(job in completed_jobs for job in jobnames)' (use-a-generator)
deploy/runtools/run_farm_deploy_managers.py:346:12: W1201: Use lazy % formatting in logging functions (logging-not-lazy)
deploy/runtools/run_farm_deploy_managers.py:352:67: R1735: Consider using {} instead of dict() (use-dict-literal)
deploy/runtools/run_farm_deploy_managers.py:365:27: R1721: Unnecessary use of a comprehension, use instance_screen_status['simdrivers'] instead. (unnecessary-comprehension)
deploy/runtools/run_farm_deploy_managers.py:383:12: W1201: Use lazy % formatting in logging functions (logging-not-lazy)
deploy/runtools/run_farm_deploy_managers.py:298:4: R0912: Too many branches (16/12) (too-many-branches)
deploy/runtools/run_farm_deploy_managers.py:55:0: R0904: Too many public methods (21/20) (too-many-public-methods)
deploy/runtools/run_farm_deploy_managers.py:418:12: C0209: Formatting a regular string which could be a f-string (consider-using-f-string)
deploy/runtools/run_farm_deploy_managers.py:438:12: E1129: Context manager 'generator' doesn't implement __enter__ and __exit__. (not-context-manager)
deploy/runtools/run_farm_deploy_managers.py:443:33: C0209: Formatting a regular string which could be a f-string (consider-using-f-string)
deploy/runtools/run_farm_deploy_managers.py:447:12: E1129: Context manager 'generator' doesn't implement __enter__ and __exit__. (not-context-manager)
deploy/runtools/run_farm_deploy_managers.py:459:12: E1129: Context manager 'generator' doesn't implement __enter__ and __exit__. (not-context-manager)
deploy/runtools/run_farm_deploy_managers.py:459:12: E1129: Context manager 'generator' doesn't implement __enter__ and __exit__. (not-context-manager)
deploy/runtools/run_farm_deploy_managers.py:466:4: C0116: Missing function or method docstring (missing-function-docstring)
deploy/runtools/run_farm_deploy_managers.py:478:4: C0116: Missing function or method docstring (missing-function-docstring)
deploy/runtools/run_farm_deploy_managers.py:492:4: C0116: Missing function or method docstring (missing-function-docstring)
deploy/runtools/run_farm_deploy_managers.py:496:37: C0209: Formatting a regular string which could be a f-string (consider-using-f-string)
deploy/runtools/run_farm_deploy_managers.py:497:28: C0209: Formatting a regular string which could be a f-string (consider-using-f-string)
deploy/runtools/run_farm_deploy_managers.py:498:20: C0209: Formatting a regular string which could be a f-string (consider-using-f-string)
deploy/runtools/run_farm_deploy_managers.py:499:28: C0209: Formatting a regular string which could be a f-string (consider-using-f-string)
deploy/runtools/run_farm_deploy_managers.py:502:37: C0209: Formatting a regular string which could be a f-string (consider-using-f-string)
deploy/runtools/run_farm_deploy_managers.py:503:28: C0209: Formatting a regular string which could be a f-string (consider-using-f-string)
deploy/runtools/run_farm_deploy_managers.py:504:20: C0209: Formatting a regular string which could be a f-string (consider-using-f-string)
deploy/runtools/run_farm_deploy_managers.py:505:28: C0209: Formatting a regular string which could be a f-string (consider-using-f-string)
deploy/runtools/run_farm_deploy_managers.py:508:4: C0116: Missing function or method docstring (missing-function-docstring)
deploy/runtools/run_farm_deploy_managers.py:514:37: C0209: Formatting a regular string which could be a f-string (consider-using-f-string)
deploy/runtools/run_farm_deploy_managers.py:515:20: C0209: Formatting a regular string which could be a f-string (consider-using-f-string)
deploy/runtools/run_farm_deploy_managers.py:525:37: C0209: Formatting a regular string which could be a f-string (consider-using-f-string)
deploy/runtools/run_farm_deploy_managers.py:526:20: C0209: Formatting a regular string which could be a f-string (consider-using-f-string)
deploy/runtools/run_farm_deploy_managers.py:530:37: C0209: Formatting a regular string which could be a f-string (consider-using-f-string)
deploy/runtools/run_farm_deploy_managers.py:531:20: C0209: Formatting a regular string which could be a f-string (consider-using-f-string)
deploy/runtools/run_farm_deploy_managers.py:534:37: C0209: Formatting a regular string which could be a f-string (consider-using-f-string)
deploy/runtools/run_farm_deploy_managers.py:535:20: C0209: Formatting a regular string which could be a f-string (consider-using-f-string)
deploy/runtools/run_farm_deploy_managers.py:618:4: W0235: Useless super delegation in method '__init__' (useless-super-delegation)
deploy/runtools/run_farm_deploy_managers.py:621:4: C0116: Missing function or method docstring (missing-function-docstring)
deploy/runtools/run_farm_deploy_managers.py:626:12: E1129: Context manager 'generator' doesn't implement __enter__ and __exit__. (not-context-manager)
deploy/runtools/run_farm_deploy_managers.py:629:21: W1514: Using open without explicitly specifying an encoding (unspecified-encoding)
deploy/runtools/run_farm_deploy_managers.py:629:45: C0103: Variable name "f" doesn't conform to snake_case naming style (invalid-name)
deploy/runtools/run_farm_deploy_managers.py:628:16: W0612: Unused variable 'collect' (unused-variable)
deploy/runtools/run_farm_deploy_managers.py:658:8: W0107: Unnecessary pass statement (unnecessary-pass)
deploy/runtools/run_farm_deploy_managers.py:11:0: C0411: standard import "import time" should be placed before "from fabric.api import prefix, local, run, env, cd, warn_only, put, settings, hide" (wrong-import-order)
deploy/runtools/run_farm_deploy_managers.py:12:0: C0411: standard import "from os.path import join as pjoin" should be placed before "from fabric.api import prefix, local, run, env, cd, warn_only, put, settings, hide" (wrong-import-order)
deploy/runtools/run_farm_deploy_managers.py:17:0: C0411: standard import "from typing import List, Dict, Optional, Union, TYPE_CHECKING" should be placed before "from fabric.api import prefix, local, run, env, cd, warn_only, put, settings, hide" (wrong-import-order)
deploy/runtools/run_farm_deploy_managers.py:11:0: W0611: Unused import time (unused-import)

------------------------------------------------------------------
Your code has been rated at 6.50/10 (previous run: 6.50/10, +0.00)

```

</details>